File: backend/utils/publicUrl.js
const ABSOLUTE_HTTP_URL_RE = /^https?:\/\//i;
const IS_PROD = String(process.env.NODE_ENV || "").toLowerCase() === "production";

function firstForwardedValue(value) {
  if (Array.isArray(value)) return firstForwardedValue(value[0]);
  return String(value || "")
    .split(",")[0]
    .trim();
}

function sanitizeProto(value) {
  const proto = firstForwardedValue(value).toLowerCase();
  return proto === "https" || proto === "http" ? proto : "";
}

function normalizeAbsoluteHttpUrl(value) {
  const raw = String(value || "").trim();
  if (!ABSOLUTE_HTTP_URL_RE.test(raw)) return "";

  try {
    const url = new URL(raw);
    if (IS_PROD) {
      url.protocol = "https:";
    }
    return url.toString().replace(/\/+$/, "");
  } catch {
    return "";
  }
}

export function isAbsoluteHttpUrl(value) {
  return ABSOLUTE_HTTP_URL_RE.test(String(value || "").trim());
}

export function getPublicBaseUrl(req) {
  const configuredBaseUrl =
    normalizeAbsoluteHttpUrl(process.env.EXTERNAL_BASE_URL) ||
    normalizeAbsoluteHttpUrl(process.env.HOST) ||
    normalizeAbsoluteHttpUrl(process.env.WEB_URL);

  if (configuredBaseUrl) return configuredBaseUrl;

  const proto =
    sanitizeProto(req.headers["x-forwarded-proto"]) || req.protocol || "http";
  const host =
    firstForwardedValue(req.headers["x-forwarded-host"]) || req.get("host") || "";

  if (!host) return "";

  return normalizeAbsoluteHttpUrl(`${proto}://${host}`);
}

export function toPublicUrl(req, value, options = {}) {
  const { absolute = true } = options;
  const raw = value == null ? "" : String(value).trim();

  if (!raw) return raw;

  if (isAbsoluteHttpUrl(raw)) {
    return normalizeAbsoluteHttpUrl(raw) || raw;
  }

  if (!absolute) return raw;

  const baseUrl = getPublicBaseUrl(req);
  if (!baseUrl) return raw;

  try {
    return new URL(raw, `${baseUrl}/`).toString();
  } catch {
    const normalizedBase = baseUrl.replace(/\/+$/, "");
    const normalizedPath = raw.replace(/^\/+/, "");
    return `${normalizedBase}/${normalizedPath}`;
  }
}

File: backend-go/internal/infra/httpx/public_url.go
package httpx

import (
	"net/http"
	"net/url"
	"os"
	"strings"

	"github.com/gin-gonic/gin"
)

func ToPublicURL(c *gin.Context, value string, absolute bool) string {
	raw := strings.TrimSpace(value)
	if raw == "" {
		return raw
	}

	if isAbsoluteHTTPURL(raw) {
		return normalizeAbsoluteHTTPURL(raw)
	}

	if !absolute {
		return raw
	}

	baseURL := publicBaseURL(c)
	if baseURL == "" {
		return raw
	}

	parsedBase, err := url.Parse(baseURL)
	if err != nil {
		return raw
	}
	parsedValue, err := url.Parse(raw)
	if err != nil {
		return raw
	}
	return parsedBase.ResolveReference(parsedValue).String()
}

func publicBaseURL(c *gin.Context) string {
	for _, key := range []string{"EXTERNAL_BASE_URL", "HOST", "WEB_URL"} {
		if value := normalizeAbsoluteHTTPURL(os.Getenv(key)); value != "" {
			return value
		}
	}

	proto := forwardedValue(c.GetHeader("X-Forwarded-Proto"))
	if proto == "" {
		if c.Request != nil && c.Request.TLS != nil {
			proto = "https"
		} else {
			proto = "http"
		}
	}

	host := forwardedValue(c.GetHeader("X-Forwarded-Host"))
	if host == "" && c.Request != nil {
		host = c.Request.Host
	}
	if host == "" {
		return ""
	}

	return normalizeAbsoluteHTTPURL(proto + "://" + host)
}

func forwardedValue(value string) string {
	return strings.TrimSpace(strings.Split(strings.TrimSpace(value), ",")[0])
}

func isAbsoluteHTTPURL(value string) bool {
	raw := strings.ToLower(strings.TrimSpace(value))
	return strings.HasPrefix(raw, "http://") || strings.HasPrefix(raw, "https://")
}

func normalizeAbsoluteHTTPURL(value string) string {
	raw := strings.TrimSpace(value)
	if !isAbsoluteHTTPURL(raw) {
		return ""
	}

	parsed, err := url.Parse(raw)
	if err != nil {
		return ""
	}
	if strings.EqualFold(strings.TrimSpace(os.Getenv("NODE_ENV")), "production") {
		parsed.Scheme = "https"
	}
	parsed.Path = strings.TrimRight(parsed.Path, "/")
	return strings.TrimRight(parsed.String(), "/")
}

func ClearCacheHeaders(w http.ResponseWriter) {
	w.Header().Del("Cache-Control")
}
